sycl/include/CL/sycl/ONEAPI/accessor_property_list.hpp

```diff
@@ -0,0 +1,232 @@
+//==----- accessor_property_list.hpp --- SYCL accessor property list -------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/access/access.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/property_list_base.hpp>
+#include <CL/sycl/property_list.hpp>
+
+#include <type_traits>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+// Forward declaration
+template <typename, int, access::mode, access::target, access::placeholder,
+          typename PropertyListT>
+class accessor;
+namespace detail {
+// This helper template must be specialized for nested instance template
+// of each compile-time-constant property.
+template <typename T> struct IsCompileTimePropertyInstance : std::false_type {};
+} // namespace detail
+namespace ONEAPI {
+
+template <typename T> struct is_compile_time_property : std::false_type {};
+
+/// Objects of the accessor_property_list class are containers for the SYCL
+/// properties.
+///
+/// Unlike \c property_list, accessor_property_list can take
+/// compile-time-constant properties.
+///
+/// \sa accessor
+/// \sa property_list
+///
+/// \ingroup sycl_api
+template <typename... PropsT>
+class accessor_property_list : protected sycl::detail::PropertyListBase {
+  // These structures check if compile-time-constant property is present in
+  // list. For runtime properties this check is always true.
+  template <class T, class U> struct AreSameTemplate : std::is_same<T, U> {};
+  template <template <class...> class T, class T1, class T2>
+  struct AreSameTemplate<T<T1>, T<T2>> : std::true_type {};
+#if __cplusplus >= 201703L
+  // Declaring non-type template parameters with auto is a C++17 feature. Since
+  // the extension is written against SYCL 2020, which implies use of C++17,
+  // there's no need to provide alternative implementations for older standards.
+  template <template <auto...> class T, auto... T1, auto... T2>
+  struct AreSameTemplate<T<T1...>, T<T2...>> : std::true_type {};
+#endif
+  // This template helps to identify if PropListT parameter pack contains
+  // property of PropT type, where PropT is a nested instance template of
+  // compile-time-constant property.
+  template <typename PropT, typename... PropListT> struct ContainsProperty;
+  template <typename PropT> struct ContainsProperty<PropT> : std::false_type {};
+  template <typename PropT, typename Head, typename... Tail>
+  struct ContainsProperty<PropT, Head, Tail...>
+      : std::conditional<
+            AreSameTemplate<PropT, Head>::value ||
+                !sycl::detail::IsCompileTimePropertyInstance<PropT>::value,
```

Comment:
Now, if the `PropT` and `Head` are not the same and `PropT` isn't a compile-time property this condition will evaluate to `true`, isn't it? Is it even correct?

Reply:
It was done to skip runtime properties, but that was probably the wrong place. Moved.